dips/dip-168.md

```diff
@@ -0,0 +1,114 @@
+
+## Acceptance and Rejection Criteria
+
+The acceptance and rejection criteria for a pair `(i, CRSN(k))` where `i` is the sequence nonce assigned to a transaction sent by `A` and `CRSN(k)` is the CRSN state under `A` at the time of processing is as follows:
+
+1. If `i` is greater than or equal to `CRSN(k).min_nonce + k` the transaction    is discarded.
+2. If `i` is less than `CRSN(k).min_nonce` the transaction is discarded.
+3. If `i` is greater than or equal to `CRSN(k).min_nonce` and less than    `CRSN(k).min_nonce + k` then
+    1.  If the slot at position `i` in `CRSN(k).slots` is `false` then `CRSN(k).slots[i]` is set to `true` and the transaction is accepted. Subsequently, the [shifting algorithm](#shifting-algorithm) is applied to `CRSN(k)`. `A`'s sequence number is incremented.
+    2. If the slot at position `i` in `CRSN(k).slots` is `true` the transaction is rejected.
+
+## Shifting algorithm
+
+Let `j = max {m | CRSN(k).min_nonce <= e <= m < k . forall CRSN(k).slots[e] == true }` (i.e., the least index in `CRSN(k).slots` such that all slots less than it are set to `true`).  The following updates are then performed:
+
+* `CRSN(k).min_nonce = CRSN(k).min_nonce + j`
+* for all `j <= i < k, ``CRSN(k).slot[i - j] = CRSN(k).slot[i]` and, for all `k - j <= i < k, ``CRSN(k).slot[i] = false`.
+
+## Sequence number to CRSN transition
+
+An account may send a transaction to switch to usage of a CRSN for transaction replay prevention. Let the sequence number of this transaction sent from account `A` be `s`, and `k` be the specified number of slots for the CRSN. A CRSN `CRSN(k)` will be published under `A` with the following states:
+
+* `CRSN(k).min_nonce = s + 1`
+* Every element of `CRSN(k).slots` is initialized to `false`.
+
+## Force Expiration
+
+An account `A` that has opted-in to using a CRSN may force expire either a single sequence nonce, or may expire a batch of sequence nonces. A sequence nonce that has been expired will cause any transactions that have been assigned that sequence nonce to be rejected according to the Acceptance and Rejection Criteria above.
+
+A singular sequence nonce may be expired by a no-op transaction assigned the sequence nonce that is to be expired. This will then perform the actions described in 3(a). Any further transactions assigned this sequence nonce will then be rejected by the Acceptance and Rejection Criteria.
+
+If a range of sequence nonces is to be expired, it can be done so by sending a specific force-expiration transaction — `force_shift(j)` — that contains the amount by which the window should be shifted `j`. This can be informally thought of as incrementing the account's `CRSN(k).min_nonce`  by `j`, and performing a "left shift" on the account's `CRSN(k).slots. `More precisely: let `CRSN(k)` be `A`'s CRSN at the time of processing the `force_shift(j)` transaction and let this transaction be sent with sequence none `m`, then the following updates will be performed (in order):
```

Comment:
IIUC, a single no-op txn that invalidates more than one sequence number will break the existing sequence number property that every account, sequence number maps to a single transaction.

I get this would be more efficient rather than sending 1 txn for every shift, but I don't know what downstream effects this could have.  Is this a concern?

Reply:
Good question. The `force_shift` operation would be a transaction itself, and it's important to note here (and I'll clarify this in this paragraph) that this force shift does not apply to the sequence number -- the `force_shift` transaction will still cause the account's sequence number to increase by one, and that's it. Because of this, it will still maintain the mapping of (account, sequence_number) mapping to a single transaction.